Intersect.Server/Entities/Player.cs

```diff
@@ -3518,88 +3518,76 @@ public void SellItem(int slot, int amount)
-                                        ), shop.SellingItems[slot].CostItemQuantity * buyItemAmt
-                                    );
+            var currencyBase = ItemBase.Get(boughtItem.CostItemId);
+            if (currencyBase == default)
+            {
+                PacketSender.SendChatMsg(this, Strings.Shops.error, ChatMessageType.Inventory, CustomColors.Alerts.Error, Name);
+                return;
+            }
+
+            var boughtItemAmount = 1;
+            if (boughtItemBase.IsStackable)
+            {
+                boughtItemAmount = Math.Max(1, amount);
+            }
 
-                                    TryGiveItem(buyItemNum, buyItemAmt);
+            var currencyAmount = FindInventoryItemQuantity(currencyBase.Id);
+            var itemCostTotal = boughtItem.CostItemQuantity * boughtItemAmount;
+            if (currencyAmount < itemCostTotal)
+            {
+                PacketSender.SendChatMsg(this, Strings.Shops.cantafford, ChatMessageType.Inventory, CustomColors.Alerts.Error, Name);
+                return;
+            }
 
-                                    if (!TextUtils.IsNone(shop.BuySound))
-                                    {
-                                        PacketSender.SendPlaySound(this, shop.BuySound);
-                                    }
-                                }
-                                else
-                                {
-                                    PacketSender.SendChatMsg(
-                                        this, Strings.Shops.inventoryfull, ChatMessageType.Inventory, CustomColors.Alerts.Error, Name
-                                    );
-                                }
-                            }
-                        }
-                        else
-                        {
-                            PacketSender.SendChatMsg(this, Strings.Shops.cantafford, ChatMessageType.Inventory, CustomColors.Alerts.Error, Name);
-                        }
+            if (!CanGiveItem(boughtItemBase.Id, boughtItemAmount))
+            {
+                PacketSender.SendChatMsg(this, Strings.Shops.inventoryfull, ChatMessageType.Inventory, CustomColors.Alerts.Error, Name);
+            }
+
+            if (itemCostTotal > 0)
+            {
+                var currencySlots = FindInventoryItemSlots(currencyBase.Id);
+                int remainingCost = itemCostTotal;
+
+                foreach (var itemSlot in currencySlots)
+                {
+                    int quantityToRemove = Math.Min(remainingCost, itemSlot.Quantity);
+                    TryTakeItem(itemSlot.ItemId, quantityToRemove);
+                    remainingCost -= quantityToRemove;
+
+                    if (remainingCost <= 0)
+                    {
+                        break;
                     }
                 }
             }
+
+            TryGiveItem(boughtItemBase.Id, boughtItemAmount);
```

Comment:
If this fails it needs to log that it failed to give the item, to what player ID, and how much was purchased.